autogpt/processing/text.py

```diff
@@ -1,170 +1,234 @@
-        ),
-    )
-    scroll_ratio = 1 / len(chunks)
-
-    for i, chunk in enumerate(chunks):
-        if driver:
-            scroll_to_percentage(driver, scroll_ratio * i)
-        logger.info(f"Adding chunk {i + 1} / {len(chunks)} to memory")
+    if instruction and question:
+        raise ValueError("Parameters 'question' and 'instructions' cannot both be set")
 
-        memory_to_add = f"Source: {url}\n" f"Raw content part#{i + 1}: {chunk}"
+    model = CFG.fast_llm_model
 
-        memory = get_memory(CFG)
-        memory.add(memory_to_add)
+    if question:
+        instruction = (
+            f'include any information that can be used to answer the question "{question}". '
+            "Do not directly answer the question itself"
+        )
 
-        messages = [create_message(chunk, question)]
-        tokens_for_chunk = count_message_tokens(messages, model)
-        logger.info(
-            f"Summarizing chunk {i + 1} / {len(chunks)} of length {len(chunk)} characters, or {tokens_for_chunk} tokens"
+    summarization_prompt = ChatPrompt.for_model(model)
+
+    token_length = count_string_tokens(text, model)
+    logger.info(f"Text length: {token_length} tokens")
+
+    # reserve 50 tokens for summary prompt, 500 for the response
+    max_chunk_length = _max_chunk_length(model) - 550
+    logger.info(f"Max chunk length: {max_chunk_length} tokens")
+
+    if not must_chunk_content(text, model, max_chunk_length):
+        # summarization_prompt.add("user", text)
+        summarization_prompt.add(
+            "user",
+            "Write a concise summary of the following text"
+            f"{f'; {instruction}' if instruction is not None else ''}:"
+            "\n\n\n"
+            f'LITERAL TEXT: """{text}"""'
+            "\n\n\n"
+            "CONCISE SUMMARY: The text is best summarized as"
+            # "Only respond with a concise summary or description of the user message."
         )
 
+        logger.debug(f"Summarizing with {model}:\n{summarization_prompt}\n")
         summary = create_chat_completion(
-            model=model,
-            messages=messages,
-        )
-        summaries.append(summary)
-        logger.info(
-            f"Added chunk {i + 1} summary to memory, of length {len(summary)} characters"
+            summarization_prompt, model, temperature=0, max_tokens=500
         )
 
-        memory_to_add = f"Source: {url}\n" f"Content summary part#{i + 1}: {summary}"
+        logger.debug(f"\n{'-'*16} SUMMARY {'-'*17}\n{summary}\n{'-'*42}\n")
+        return summary.strip(), None
 
-        memory.add(memory_to_add)
+    summaries: list[str] = []
+    chunks = list(split_text(text, for_model=model, max_chunk_length=max_chunk_length))
+
+    for i, (chunk, chunk_length) in enumerate(chunks):
+        logger.info(
+            f"Summarizing chunk {i + 1} / {len(chunks)} of length {chunk_length} tokens"
+        )
+        summary, _ = summarize_text(chunk, instruction)
```

Comment:
Recursively invoking a function that makes language model calls seems like a **very bad** and potentially extremely expensive idea.

Reply:
Strongly suggest we make this function iteratiive instead of recursive and determine how many summarization calls we'll make before we do (and just error if it's more than like 10 or 50 or something sensible).

Reply:
Because of the way it works, it should never recurse more than 2 levels, and that should already be rare:

How it works:
1. It chunks and summarizes all chunks (of 3500+ tokens) into short summaries (rarely >100 tokens, so a factor ~35 reduction of content length). It chunks linearly and will not recurse more than 1 level at this stage.
2. It combines all the summaries into one grand total summary. It will recurse 1 level here, but can recurse further if either the partial summaries are very long (not sure if that ever happens) or if there are very many chunk summaries, so that their combined length exceeds the maximum length of the context window. This, too, should rarely ever happen.

Reply:
I can see what you mean that the recursion is limited. The main issue here is the number of api calls and not the recursion, which I should have framed more clearly.  This is fine for now.  

At your discretion: maybe bail if 
```
chunks = list(split_text(text, for_model=model, max_chunk_length=max_chunk_length))
```
gives you more than some reasonable number of chunks.  I think the edge cases (summarize a whole book, e.g.) will happen and people will complain.  

You could add a config key for a summarization limit or something or just add a hard cap.

Reply:
Will do this later